core/tx_list.go

```diff
@@ -340,3 +343,205 @@ func (l *txList) Empty() bool {
+type priceHeap []*big.Int
+
+func (h priceHeap) Len() int           { return len(h) }
+func (h priceHeap) Less(i, j int) bool { return h[i].Cmp(h[j]) < 0 }
+func (h priceHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
+
+func (h *priceHeap) Push(x interface{}) {
+	*h = append(*h, x.(*big.Int))
+}
+
+func (h *priceHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+// priceKey is the hash map key type representing an unsigned big integer.
+type priceKey [256 / int(unsafe.Sizeof(uintptr(0)))]big.Word
+
+// newPriceKey converts a bit integer to a hash map key.
+func newPriceKey(x *big.Int) priceKey {
+	var key priceKey
+	copy(key[:], x.Bits())
+	return key
+}
+
+// txPricedMap is a price->transactions hash map with a heap based index to allow
+// iterating over the contents in a price-incrementing way.
+type txPricedMap struct {
+	items  map[priceKey]map[common.Hash]*types.Transaction // Hash map storing the transaction data
+	index  *priceHeap                                      // Heap of prices of all the stored transactions
+	stales int                                             // Number of stale price points to (re-heap trigger)
+	logger log.Logger                                      // Logger reporting basic price pool stats
+}
+
+// newTxPricedMap creates a new price-sorted transaction map.
+func newTxPricedMap() *txPricedMap {
+	m := &txPricedMap{
+		items: make(map[priceKey]map[common.Hash]*types.Transaction),
+		index: new(priceHeap),
+	}
+	m.logger = log.New(
+		"prices", log.Lazy{Fn: func() int { return len(*m.index) - m.stales }},
+		"stales", log.Lazy{Fn: func() int { return m.stales }},
+	)
+	return m
+}
+
+// Put inserts a new transaction into the map, also updating the map's price index.
+func (m *txPricedMap) Put(tx *types.Transaction) {
+	price, hash := tx.GasPrice(), tx.Hash()
+
+	// Generate the key and ensure we have a valid map for it
+	key := newPriceKey(price)
+	if m.items[key] == nil {
+		m.items[key] = make(map[common.Hash]*types.Transaction)
+		heap.Push(m.index, price)
+	} else if len(m.items[key]) == 0 {
+		m.stales--
+	}
+	// Add the transaction to the map
+	m.items[key][hash] = tx
+
+	m.logger.Trace("Accepted new transaction", "hash", hash, "price", price)
+}
+
+// Remove looks up a transaction and deletes it from the sorted map. Even if no
+// more transactions are left with the same price point, the price point itself
+// is retained to achieve hysteresis, until a staleness threshold is reached.
+func (m *txPricedMap) Remove(tx *types.Transaction) {
+	price, hash := tx.GasPrice(), tx.Hash()
+
+	key := newPriceKey(price)
+	if txs := m.items[key]; txs != nil {
+		if txs[hash] != nil {
+			// Transaction found, delete and update stale counter
+			delete(m.items[key], hash)
+			if len(m.items[key]) == 0 {
+				m.stales++
+			}
+			// If the number of stale entries reached a critical threshold (10%), re-heap and clean up
```

Comment:
I bumped the percentage to 25%, but no. Stale means that the entire price point got empty (i.e all th transactions from that point were removed). If I have 1K txs spread over 10 levels, it means I need to remove at least 100tx for the first stale point to appear.